### PR TITLE
Migrate to Coil 3

### PR DESCRIPTION
**Pull request by Isira-Seneviratne**
#### What is it?
- [ ] Bugfix (user facing)
- [ ] Feature (user facing)
- [x] Codebase improvement (dev facing)
- [x] Meta improvement to the project (dev facing)

#### Description of the changes in your PR

- Update Coil from 2.7.0 to 3.0.3 ([changelog](https://github.com/coil-kt/coil/blob/main/CHANGELOG.md)).
- Convert the application class to Kotlin due to some methods in `ImageLoader.Builder` being converted to extension functions.
- Update the codebase to account for the changes introduced in Coil 3.

#### Fixes the following issue(s)

- Fixes #

#### APK testing

The APK can be found by going to the "Checks" tab below the title. On the left pane, click on "CI", scroll down to "artifacts" and click "app" to download the zip file which contains the debug APK of this PR. You can find more info and a video demonstration [on this wiki page](https://github.com/TeamNewPipe/NewPipe/wiki/Download-APK-for-PR).

#### Due diligence
- [x] I read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md).

### CODE REVIEW

**Review by Profpatsch — APPROVED** [review]
Tried it out, app still works the same for thumbnails, code looks :ok:

**Comment by Profpatsch** [comment]
I think you have to run the auto-formatter on the .kt file, otherwise the linter fails

File: app/src/main/java/org/schabi/newpipe/App.kt [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+                fun isThrowableCritical(throwable: Throwable): Boolean {
+                    // Though these exceptions cannot be ignored
+                    return throwable
+                        .hasAssignableCause(
+                            NullPointerException::class.java,
+                            IllegalArgumentException::class.java, // bug in app
+                            OnErrorNotImplementedException::class.java,
+                            MissingBackpressureException::class.java,
+                            IllegalStateException::class.java,
+                        ) // bug in operator
```
Comment (Stypox) [inline app/src/main/java/org/schabi/newpipe/App.kt:193]:
```suggestion
                        .hasAssignableCause(
                            // bug in app
                            NullPointerException::class.java,
                            IllegalArgumentException::class.java,
                            OnErrorNotImplementedException::class.java,
                            MissingBackpressureException::class.java,
                            // bug in operator
                            IllegalStateException::class.java,
                        )
```

File: app/src/main/java/org/schabi/newpipe/settings/ContentSettingsFragment.java [hunk 2]
```diff
@@ -41,7 +41,7 @@ public void onCreatePreferences(final Bundle savedInstanceState, final String ro
                 (preference, newValue) -> {
                     ImageStrategy.setPreferredImageQuality(PreferredImageQuality
                             .fromPreferenceKey(requireContext(), (String) newValue));
-                    final var loader = Coil.imageLoader(preference.getContext());
+                    final var loader =  SingletonImageLoader.get(preference.getContext());
```
Comment (Stypox) [inline app/src/main/java/org/schabi/newpipe/settings/ContentSettingsFragment.java:44]:
```suggestion
                    final var loader = SingletonImageLoader.get(preference.getContext());
```

**Review by Stypox — APPROVED** [review]
I tested the following things and everything worked well:
- clearing cache actually removes the files on disk
- the player thumbnail loads normally (also in SoundCloud songs)
- the notification thumbnail also loads fine
- changing image quality works
- sharing a video also shows its image in the share sheet, so the image is correctly obtained from cache

Thank you!

**Pull request merged**